packages/ui/src/components/Channel/Scroller.tsx

```diff
@@ -717,3 +634,205 @@ const PressableMessage = React.memo(
+
+      setNeedsScrollToAnchor(false);
+
+      logger.log('doing a scrollToIndex to anchor', {
+        anchorIndex,
+        readyToDisplayPosts,
+      });
+
+      // If we aren't showing the posts yet, we can jump without animation.
+      // Once we show the posts, use an animation for the poor little user's eyesies
+      const shouldAnimateScroll = readyToDisplayPosts;
+      flatListRef.current.scrollToIndex({
+        index: anchorIndex,
+        animated: shouldAnimateScroll,
+        viewPosition: anchor?.type === 'unread' ? 1 : 0.5,
+      });
+    } finally {
+      logger.log('finished scroll attempt');
+      isScrollAttemptActiveRef.current = false;
+    }
+  }, [
+    didAnchorSearchTimeout,
+    needsScrollToAnchor,
+    anchorIndex,
+    anchor?.type,
+    userHasScrolled,
+    readyToDisplayPosts,
+    flatListRef,
+  ]);
+
+  const handleScrollBeginDrag = useCallback(() => {
+    setUserHasScrolled(true);
+  }, []);
+
+  // We use this function to manage autoscrolling to the anchor post. We need
+  // the post to be rendered before we're able to scroll to it, so we wait for
+  // it here. Once it's rendered we scroll to it and set `needsScrollToAnchor` to
+  // false, revealing the Scroller.
+  const handleItemLayout = useCallback(
+    (post: db.Post, index: number) => {
+      renderedPostsRef.current.add(post.id);
+
+      if (didAnchorSearchTimeout) {
+        return;
+      }
+
+      // If the anchor post got a layout, attempt a scroll.
+      if (post.id === anchor?.postId) {
+        logger.log('scrolling to initially set anchor', post.id, index);
+        scrollToAnchorIfNeeded();
+      }
+
+      // If we've got at least a page of posts and we've rendered them all,
+      // reveal the scroller to prevent getting stuck when messages are
+      // deleted.
+      if (posts?.length && renderedPostsRef.current.size >= posts?.length) {
+        setDidAnchorSearchTimeout(true);
+      }
+    },
+    [
+      anchor?.postId,
+      posts?.length,
+      scrollToAnchorIfNeeded,
+      didAnchorSearchTimeout,
+    ]
+  );
+  const maintainVisibleContentPositionConfig = useMemo(() => {
+    return channelType === 'chat' ||
+      channelType === 'dm' ||
+      channelType === 'groupDm'
+      ? {
+          minIndexForVisible: 0,
+          // If this is set to a number, the list will scroll to the bottom (or top,
+          // if not inverted) when the list height changes. This is undesirable when
+          // we're starting at an older post and scrolling down towards newer ones,
+          // as it will trigger on every new page load, causing jumping. Instead, we
+          // only enable it when there's nothing newer left to load (so, for new incoming messages only).
+          autoscrollToTopThreshold:
+            !userHasScrolled || hasNewerPosts ? undefined : 0,
+        }
+      : undefined;
+  }, [userHasScrolled, hasNewerPosts, channelType]);
+
+  const handleScrollToIndexFailed = useCallback(
```

Comment:
in theory, should this only ever be called once per unique anchor?

Reply:
I *think* there is an edge case where this is called multiple times:
1. Attempt to scroll to anchor index
2. Scroll to index fails, causing a "best effort" scrollToOffset
3. We later think we can scroll correctly, so we do scrollToAnchorIfNeeded
4. but that scroll to index fails again, looping back to step 2

I'm not sure if this would happen in practice (one hypothetical situation is if extra list content gets loaded in between the second invocation of scrollToAnchorIfNeeded and the actual scrollToIndex which changes the layout of the anchor)